src/helpers/get-log.ts

```diff
@@ -14,17 +14,22 @@
  * app.log.fatal("Goodbye, cruel world!");
  * ```
  */
-import pino, { LoggerOptions } from "pino";
+import pino, { Logger, LoggerOptions } from "pino";
 import { getTransformStream, Options, LogLevel } from "@probot/pino";
+import { isProd } from "./is-prod";
 
 export type GetLogOptions = {
   level?: LogLevel;
   logMessageKey?: string;
 } & Options;
 
-export function getLog(options: GetLogOptions = {}) {
+export function getLog(options: GetLogOptions = {}): Logger {
   const { level, logMessageKey, ...getTransformStreamOptions } = options;
 
+  if (isProd() && !getTransformStreamOptions.logFormat) {
```

Comment:
I think we shouldn't access `process.env` here. `isProduction` should instead be an option that you pass to as `options`. `getTransformStreamOptions.logFormat` should be set instead when calling `getLog()`

Reply:
I'm not sure I follow the recommendation here. If the `logFormat` should be set when calling `getLog`, why does the `isProduction` have to be passed as `options`?

Did you mean I should follow either of these approaches?

I have looked at the call-sites of `getLog` and I fear it could be difficult to ensure `logFormat` is set. I'd have to set the `logFormat` in various places and I don't think that makes the code nicer. With the current approach, the decision is at least in one place.

Reply:
### TL;DR

I had a look at the code and we don't need to add a `isProduction` option here. I don't even think we need the `isProduction()` helper method. Instead, set a default for `logFormat` [here](https://github.com/probot/probot/blob/02a57d1bd2948ed019d5046c799f7e6fae770fc4/src/bin/read-env-options.ts#L19) based on `process.env.NODE_ENV`

### The longer version

Interacting with environment variables is side effect. That's what we are trying to avoid. We cannot be sure that `process.env` is always set, it depends on the runtime environment. Accessing `process.env` is a convenient shortcut but it's hard to test and causes hard to debug issues down the road. The Probot codebase used to access `process.env` all over the place until a recent refactoring, I don't want to go back there.

We access `process.env` in the entrypoints that are meant to be called from the terminal. Many of the modules can be imported as libraries though, such as 

```js
import { Probot } from "probot"
```

The `Probot` class is using the `getLog` method in its constructor [here](https://github.com/probot/probot/blob/02a57d1bd2948ed019d5046c799f7e6fae770fc4/src/probot.ts#L54). So we want to avoid accessing any environment variables.

When you run `probot run ./app.js`, then we create a custom instance of the pino logger [here](https://github.com/probot/probot/blob/02a57d1bd2948ed019d5046c799f7e6fae770fc4/src/run.ts#L65), `logFormat` is already set based on the environment and CLI option: https://github.com/probot/probot/blob/02a57d1bd2948ed019d5046c799f7e6fae770fc4/src/run.ts#L35. The environment variables are read here:
https://github.com/probot/probot/blob/02a57d1bd2948ed019d5046c799f7e6fae770fc4/src/bin/read-env-options.ts

This is the place where we explicitly interact with environment variables and where I would set a default for `logFormat` base on `process.env.NODE_ENV`. That should make it work the way you want. The `Server` and `Probot` classes and `createProbot()` use the `getLog` method, but only if `options.log` is not set. And `options.log` is set when you run `probot run ./app.js` for a long living node process or use `createProbot()` in a serverless environment.

Reply:
Thanks for the explanation, @gr2m. This makes sense to me. I have addressed it [here](https://github.com/probot/probot/pull/1598/commits/cdffec03a2ff802173eb74c996e8923bded84ca8).

I have tried adding a test to run.test.ts, but I couldn't figure out how I could make sure the logging is done with json. Adding a test for `readEnvOptions` seemed a bit superficial. I'll be happy to add a test if you could give me a hint on how to approach it.

I have left the helper function `isProduction` in the code as I also used it in other places. I'll be happy to remove it and revert its usages if you'd rather want the usages of `process.env` to be explicit.

Reply:
> Thanks for the explanation, @gr2m. This makes sense to me. I have addressed it here.

Thanks!

> I have left the helper function isProduction in the code as I also used it in other places. I'll be happy to remove it and revert its usages if you'd rather want the usages of process.env to be explicit.

That's fine we can leave that in

> I'll be happy to add a test if you could give me a hint on how to approach it.

I'll get back to you, hopefully later today. I need to have a look myself, it's been a while I worked on the tests.

Reply:
I added a test here:
https://github.com/probot/probot/pull/1598/commits/06b33c5e0b5ec2c73c2ef25991df862529e461f4

I had to change your implementation slightly to make it test-able
https://github.com/probot/probot/pull/1598/commits/0ca219f8e8b17ec98336b7956721ac36eeb09853

Let me know if that looks good to you

Reply:
ha! there's a helper to capture log output 😄 

Very nice, LGTM 👍